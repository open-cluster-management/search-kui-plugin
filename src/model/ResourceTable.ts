/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project

export interface TableProps {
    items: object[],
    kind: string,
  }

export interface TableState {
  itemForAction: object,
  page: number,
  pageSize: number,
  sortDirection: any,
  selectedKey: string,
  modalOpen: boolean,
  collapse: boolean,
  action: string
}
