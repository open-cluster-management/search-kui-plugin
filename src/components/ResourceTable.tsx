/* eslint-disable react/no-deprecated */
/* eslint-disable no-case-declarations */
/*******************************************************************************
* Licensed Materials - Property of IBM
* (c) Copyright IBM Corporation 2019. All Rights Reserved.
*
* Note to U.S. Government Users Restricted Rights:
* Use, duplication or disclosure restricted by GSA ADP Schedule
* Contract with IBM Corp.
*******************************************************************************/

// Hack to workaround build issues with Carbon dependencies
if (!window || !window.navigator || !window.navigator.userAgent) {
  Object.defineProperty(window, 'navigator', { value: { userAgent: 'node'}, writable: true })
  Object.defineProperty(document, 'getElementById', { value: (val: string) => document.querySelector('#' + val), writable: true })
}

import * as React from 'react'
import * as lodash from 'lodash'
import * as PropTypes from 'prop-types'
import tableDefinitions from '../definitions/search-definitions'
import Modal from './Modal'
import { Pagination, DataTable } from 'carbon-components-react'
import { TableProps, TableState } from '../model/ResourceTable'
import { getCurrentTab } from '@kui-shell/core'
import strings from '../util/i18n'
import { Delete16, Edit16, Share16 } from '@carbon/icons-react'
import { getStatusIcon } from '../util/status'

const { Table, TableHead, TableRow, TableBody, TableCell } = DataTable
const PAGE_SIZES = { DEFAULT: 10, VALUES: [5, 10, 20, 50, 75, 100] }

export default class ResourceTable extends React.PureComponent<TableProps, TableState> {
  static propTypes = {
    collapseTable: PropTypes.bool,
    expandFullPage: PropTypes.bool,
    items: PropTypes.array,
    kind: PropTypes.string,
  }

  tabHeaders = ['name', 'action', 'share', 'edit']

  constructor(props) {
    super(props)
    this.state = {
      action: '',
      itemForAction: {},
      page: 1,
      pageSize: PAGE_SIZES.DEFAULT,
      sortDirection: 'asc',
      selectedKey: '',
      modalOpen: false,
      collapse: false,
    }

    this.getHeaders = this.getHeaders.bind(this)
    this.getRows = this.getRows.bind(this)
    this.handleEvent = this.handleEvent.bind(this)
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      collapse: nextProps.collapseTable,
    })
  }

  toggleCollapseTable = () => {
    this.setState(({collapse}) => {
      return { collapse: !collapse }
    })
  }

  getHeaders() {
    const { kind } = this.props
    const resource = tableDefinitions[kind] || tableDefinitions['genericresource']
    const headers = resource.columns.map((col) => ({
      key: col.key, header: strings(`table.header.${col.msgKey || col.key}`),
    }))
    headers.push({ key: 'action', header: ''})

    if (kind === 'savedSearches') {
      headers.push({ key: 'edit', header: '' })
      headers.push({ key: 'share', header: '' })
    }

    return headers
  }

  getRows() {
    let { items } = this.props
    const { kind } = this.props
    let transforms = tableDefinitions[kind] || tableDefinitions['genericresource']
    const { page, pageSize, selectedKey, sortDirection } = this.state

    if (selectedKey) {
      items = lodash.orderBy(items, [selectedKey], [sortDirection])
    }
    const startItem = (page - 1) * pageSize
    const visibleItems = items.slice(startItem, startItem + pageSize)
    // Get all the transforms for current kind defined in search-definitions
    transforms = transforms.columns.filter((rowCol) => rowCol.transform)

    return visibleItems.map((item, i) => {
      const row = { id: i.toString(), action: null, ...item }

      // Transform each key that has a transform and assign it back to the row item
      if (transforms.length > 0) {
        transforms.forEach((transform) => {
          // NOTE - this will only work for getAge transfrom currently
          row[transform.key] = transform.transform(item, null, transform.key)
        })
      }

      if (this.props.kind !== 'cluster' && this.props.kind !== 'release') {
        row.action = (
          <Delete16
            className='table-action'
            onClick={() => this.setState({ itemForAction: item, modalOpen: true, action: 'remove' })}
          />
        )

        if (kind === 'savedSearches') {
          row['share'] = (
            <Share16
              className='table-action'
              onClick={() => this.setState({ itemForAction: item, modalOpen: true, action: 'share' })}
            />
          )

          row['edit'] = (
            <Edit16
              className='table-action'
              onClick={() => this.setState({ itemForAction: item, modalOpen: true, action: 'edit' })}
            />
          )
        }
      }
      return row
    })
  }

  handleSort = (selectedKey) => () => {
    if (selectedKey) {
      this.setState((preState) => {
        return {selectedKey, sortDirection: preState.sortDirection === 'asc' ? 'desc' : 'asc' }
      })
    }
  }

  handleEvent = (row, cell, e?) => {
    if ((e && e.which === 13) || !e) {
      const item = this.props.items.filter((data) => data['name'] === lodash.get(row, '[0].value', ''))

      switch (cell.info['header']) {
        case 'action':
          this.setState({ itemForAction: lodash.get(item, '[0]', ''), modalOpen: true, action: 'remove' })
          break
        case 'edit':
          this.setState({ itemForAction: lodash.get(item, '[0]', ''), modalOpen: true, action: 'edit' })
          break
        case 'share':
          this.setState({ itemForAction: lodash.get(item, '[0]', ''), modalOpen: true, action: 'share' })
          break
        default:
          const headers = ['name', 'namespace', 'cluster'] // Headers used for filters keywords
          let filters = '' // Query filters

          // Since multiple resources can have the same name, get the namespace and cluster name for that resource
          const _ = row.filter((data) => (data && data.value && (headers.includes(data.info['header']))))

          _.forEach((data) => {
            filters += `${data.info['header']}:${data.value} `
          })

          if (this.props.kind === 'savedSearches' && cell.info['header'] === 'name') {
            // When user clicks on saved search name we want to run the query seen in search text column
            return getCurrentTab().REPL.pexec(`search ${row[2].value}`)
          } else if (cell.info['header'] === 'name' && filters) {
            return getCurrentTab().REPL.pexec(`search summary kind:${this.props.kind} ${filters}`)
          } else {
            return null
          }
      }
    }
  }

  render() {
    const { page, pageSize, sortDirection, selectedKey, modalOpen, collapse } = this.state
    const totalItems = this.props.items.length
    const sortColumn = selectedKey

    return (
      <React.Fragment>
        <div className={'search--resource-table-header'}>
          <div>
            <button
              onClick={this.toggleCollapseTable}
              className={'search--resource-table-header-button'}>
              {<span className={'linked-resources'}>{`${this.props.kind}(${this.props.items.length})`}</span>}
              {!collapse ? <span className={'arrow-up'}>&#9650;</span> : <span className={'arrow-down'}>&#9660;</span>}
            </button>
          </div>
        </div>
        {!collapse
      ? <React.Fragment>
        <DataTable
          key={`${this.props.kind}-resource-table`}
          rows={this.getRows()}
          headers={this.getHeaders()}
          render={({ rows, headers }) => {
            return (
              <Table>
                <TableHead>
                  <TableRow>
                    {headers.map((header) => (
                      <th scope={'col'} key={header.key}>
                        {header.key !== 'action'
                          ? <div
                              onClick={this.handleSort(header.key)}
                              className={`bx--table-sort-v2${sortDirection === 'asc' ? ' bx--table-sort-v2--ascending' : ''}${sortColumn === header.key ? ' bx--table-sort-v2--active' : ''}`}
                              data-key={header.key} >
                              <span className='bx--table-header-label'>{header.header}
                                <span className={'arrow-header-label'}>{this.state.sortDirection === 'asc'
                                  ? <span>&#9650;</span>
                                  : <span>&#9660;</span>
                                }</span>
                              </span>
                            </div>
                          : null
                        }
                      </th>
                    ))}
                  </TableRow>
                </TableHead>
                <TableBody>
                  {rows.map((row) => (
                    <TableRow key={row.id} className='bx--data-table--compact'>
                      {row.cells.map((cell) => <TableCell
                        onKeyPress={(e) => this.handleEvent(row.cells, cell, e)}
                        tabIndex={this.tabHeaders.includes(cell.info['header']) ? 0 : null}
                        key={cell.id}
                        onClick={() => this.handleEvent(row.cells, cell) }
                        >
                      {
                        cell.info['header'] !== 'status'
                        ? cell.value
                        : <div>
                            { // If the resource contains a status, add a status icon to that column in the table.
                              getStatusIcon(cell.value)
                            }
                            <span className={`status-name`}>{`${cell.value}`}</span>
                          </div>
                      }
                      </TableCell>)}
                    </TableRow>
                  ))}
                </TableBody>
              </Table>
            )}
          }
        />

        {
          this.props.items.length > PAGE_SIZES.DEFAULT
          ? <Pagination
              key='resource-table-pagination'
              id='resource-table-pagination'
              onChange={(pagination) => this.setState(pagination)}
              pageSize={pageSize}
              pageSizes={PAGE_SIZES.VALUES}
              totalItems={totalItems}
              page={page}
              // disabled={pageSize >= totalItems}
              isLastPage={pageSize >= totalItems}
              itemsPerPageText={strings('pagination.itemsPerPage')}
              pageRangeText={(current, total) => strings('pagination.pageRange', [current, total])}
              itemRangeText={(min, max, total) => `${strings('pagination.itemRange', [min, max])} ${strings('pagination.itemRangeDescription', [total])}`}
              pageInputDisabled={pageSize >= totalItems}
            />
          : null
        }

        </React.Fragment>
        : null }

        <Modal
          item={this.state.itemForAction}
          modalOpen={modalOpen}
          onClose={() => this.setState({ itemForAction: {}, modalOpen: false, action: '' })}
          action={this.state.action}
        />
      </React.Fragment>
    )
  }
}