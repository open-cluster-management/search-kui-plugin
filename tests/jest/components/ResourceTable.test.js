/*******************************************************************************
* Licensed Materials - Property of IBM
* (c) Copyright IBM Corporation 2019. All Rights Reserved.
*
* Note to U.S. Government Users Restricted Rights:
* Use, duplication or disclosure restricted by GSA ADP Schedule
* Contract with IBM Corp.
*******************************************************************************/
'use strict'

import React from 'react'
import ResourceTable from '../../../dist/components/ResourceTable'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'

const resource = require('../../data/ResourceTableMockData')

afterEach(() => {
  // console.log('Clearing mock data')
  const mockFn = jest.fn()
  mockFn.mockReset()
})

describe('ResourceTable component', () => {
  resource.data.related.forEach(res => { // Using related data for bigger testing sample
    const WRAPPER = shallow(<ResourceTable items={res.items} kind={res.kind} />)

    const TABLE = WRAPPER.find('DataTable')
    const HEADER = { button: WRAPPER.find('button.search--resource-table-header-button') }

    HEADER['text'] = HEADER.button.text()
    HEADER['caret'] = HEADER.button.find('span.arrow-up')
    HEADER['filter'] = TABLE.find('div.bx--table-sort-v2')

    WRAPPER.instance().sortIds = jest.fn()
    WRAPPER.update()
    // console.log('instance', WRAPPER.instance())
    WRAPPER.instance().componentDidMount()

    // console.log('Headers', TABLE.props().headers.render())
    // console.log('Rows', TABLE.props().rows)
    // console.log('State', WRAPPER.state())
    // console.log('Props', TABLE.props())
    describe('mounting', () => {
      it('should mount the ResourceTable component', () => {
        expect(true).toBeTruthy()
      })
    })

    describe('data table', () => {
      it(`should render (${res.kind}) data table`, () => {
        expect(TABLE).toBeDefined()
        expect(TABLE).toHaveLength(1)
      })

      describe('onClick event', () => {
        describe('header event', () => {
          it(`should sort (${res.kind}) resources`, () => {
            console.log(TABLE)
            // expect(TABLE.props().filter('name')).toMatchSnapshot()
          })
        })
      })
    })
    
    describe('header button', () => { // Simulate Resource Table collapse
      it(`should render (${res.kind}) header button`, () => {
        expect(HEADER.button).toBeDefined()
        expect(HEADER.button).toHaveLength(1)
      })

      describe('onClick event', () => {
        it(`should collapse (${res.kind}) resource table`, () => {
          HEADER.button.simulate('click') // Click on header button
          expect(HEADER.text).toBe(`${res.kind}(${res.count})`.concat(HEADER.caret.text())) // Header count should remain the same
          expect(shallowToJson(WRAPPER)).toMatchSnapshot()
        })
  
        it(`should display (${res.kind}) resource table`, () => {
          HEADER.button.simulate('click') // Click on header button
          expect(HEADER.text).toBe(`${res.kind}(${res.count})`.concat(HEADER.caret.text())) // Header count should remain the same
          expect(shallowToJson(WRAPPER)).toMatchSnapshot()
        })
      })
    })
  })
})