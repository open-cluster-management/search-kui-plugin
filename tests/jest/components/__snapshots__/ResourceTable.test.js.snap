// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceTable component header button onClick event should collapse (cluster) resource table 1`] = `
<Fragment>
  <div
    className="search--resource-table-header"
  >
    <div>
      <button
        className="search--resource-table-header-button"
        onClick={[Function]}
      >
        <span
          className="linked-resources"
        >
          cluster(2)
        </span>
        <span
          className="arrow-down"
        >
          ▼
        </span>
      </button>
    </div>
  </div>
  <ResourceModal
    action=""
    item={Object {}}
    modalOpen={false}
    onClose={[Function]}
  />
</Fragment>
`;

exports[`ResourceTable component header button onClick event should collapse (deployment) resource table 1`] = `
<Fragment>
  <div
    className="search--resource-table-header"
  >
    <div>
      <button
        className="search--resource-table-header-button"
        onClick={[Function]}
      >
        <span
          className="linked-resources"
        >
          deployment(2)
        </span>
        <span
          className="arrow-down"
        >
          ▼
        </span>
      </button>
    </div>
  </div>
  <ResourceModal
    action=""
    item={Object {}}
    modalOpen={false}
    onClose={[Function]}
  />
</Fragment>
`;

exports[`ResourceTable component header button onClick event should collapse (node) resource table 1`] = `
<Fragment>
  <div
    className="search--resource-table-header"
  >
    <div>
      <button
        className="search--resource-table-header-button"
        onClick={[Function]}
      >
        <span
          className="linked-resources"
        >
          node(2)
        </span>
        <span
          className="arrow-down"
        >
          ▼
        </span>
      </button>
    </div>
  </div>
  <ResourceModal
    action=""
    item={Object {}}
    modalOpen={false}
    onClose={[Function]}
  />
</Fragment>
`;

exports[`ResourceTable component header button onClick event should collapse (release) resource table 1`] = `
<Fragment>
  <div
    className="search--resource-table-header"
  >
    <div>
      <button
        className="search--resource-table-header-button"
        onClick={[Function]}
      >
        <span
          className="linked-resources"
        >
          release(2)
        </span>
        <span
          className="arrow-down"
        >
          ▼
        </span>
      </button>
    </div>
  </div>
  <ResourceModal
    action=""
    item={Object {}}
    modalOpen={false}
    onClose={[Function]}
  />
</Fragment>
`;

exports[`ResourceTable component header button onClick event should collapse (replicaset) resource table 1`] = `
<Fragment>
  <div
    className="search--resource-table-header"
  >
    <div>
      <button
        className="search--resource-table-header-button"
        onClick={[Function]}
      >
        <span
          className="linked-resources"
        >
          replicaset(2)
        </span>
        <span
          className="arrow-down"
        >
          ▼
        </span>
      </button>
    </div>
  </div>
  <ResourceModal
    action=""
    item={Object {}}
    modalOpen={false}
    onClose={[Function]}
  />
</Fragment>
`;

exports[`ResourceTable component header button onClick event should collapse (secret) resource table 1`] = `
<Fragment>
  <div
    className="search--resource-table-header"
  >
    <div>
      <button
        className="search--resource-table-header-button"
        onClick={[Function]}
      >
        <span
          className="linked-resources"
        >
          secret(2)
        </span>
        <span
          className="arrow-down"
        >
          ▼
        </span>
      </button>
    </div>
  </div>
  <ResourceModal
    action=""
    item={Object {}}
    modalOpen={false}
    onClose={[Function]}
  />
</Fragment>
`;

exports[`ResourceTable component header button onClick event should collapse (service) resource table 1`] = `
<Fragment>
  <div
    className="search--resource-table-header"
  >
    <div>
      <button
        className="search--resource-table-header-button"
        onClick={[Function]}
      >
        <span
          className="linked-resources"
        >
          service(2)
        </span>
        <span
          className="arrow-down"
        >
          ▼
        </span>
      </button>
    </div>
  </div>
  <ResourceModal
    action=""
    item={Object {}}
    modalOpen={false}
    onClose={[Function]}
  />
</Fragment>
`;

exports[`ResourceTable component header button onClick event should display (cluster) resource table 1`] = `
<Fragment>
  <div
    className="search--resource-table-header"
  >
    <div>
      <button
        className="search--resource-table-header-button"
        onClick={[Function]}
      >
        <span
          className="linked-resources"
        >
          cluster(2)
        </span>
        <span
          className="arrow-up"
        >
          ▲
        </span>
      </button>
    </div>
  </div>
  <DataTable
    filterRows={[Function]}
    headers={
      Array [
        Object {
          "header": "Name",
          "key": "name",
        },
        Object {
          "header": "Namespace",
          "key": "namespace",
        },
        Object {
          "header": "Nodes",
          "key": "nodes",
        },
        Object {
          "header": "Klusterlet Version",
          "key": "klusterletVersion",
        },
        Object {
          "header": "Kubernetes Version",
          "key": "kubernetesVersion",
        },
        Object {
          "header": "CPU",
          "key": "cpu",
        },
        Object {
          "header": "Storage",
          "key": "storage",
        },
        Object {
          "header": "Memory",
          "key": "memory",
        },
        Object {
          "header": "",
          "key": "action",
        },
      ]
    }
    key="cluster-resource-table"
    locale="en"
    render={[Function]}
    rows={
      Array [
        Object {
          "action": null,
          "id": "0",
          "kind": "cluster",
          "name": "mock-cluster-1",
        },
        Object {
          "action": null,
          "id": "1",
          "kind": "cluster",
          "name": "mock-cluster=2",
        },
      ]
    }
    sortRow={[Function]}
    translateWithId={[Function]}
  />
  <ResourceModal
    action=""
    item={Object {}}
    modalOpen={false}
    onClose={[Function]}
  />
</Fragment>
`;

exports[`ResourceTable component header button onClick event should display (deployment) resource table 1`] = `
<Fragment>
  <div
    className="search--resource-table-header"
  >
    <div>
      <button
        className="search--resource-table-header-button"
        onClick={[Function]}
      >
        <span
          className="linked-resources"
        >
          deployment(2)
        </span>
        <span
          className="arrow-up"
        >
          ▲
        </span>
      </button>
    </div>
  </div>
  <DataTable
    filterRows={[Function]}
    headers={
      Array [
        Object {
          "header": "Name",
          "key": "name",
        },
        Object {
          "header": "Namespace",
          "key": "namespace",
        },
        Object {
          "header": "Cluster",
          "key": "cluster",
        },
        Object {
          "header": "Desired",
          "key": "desired",
        },
        Object {
          "header": "Current",
          "key": "current",
        },
        Object {
          "header": "Ready",
          "key": "ready",
        },
        Object {
          "header": "Available",
          "key": "available",
        },
        Object {
          "header": "Created",
          "key": "created",
        },
        Object {
          "header": "",
          "key": "action",
        },
      ]
    }
    key="deployment-resource-table"
    locale="en"
    render={[Function]}
    rows={
      Array [
        Object {
          "action": <ForwardRef(Delete16)
            className="table-action"
            onClick={[Function]}
          />,
          "apigroup": "apps",
          "available": 1,
          "cluster": "mock-cluster-1",
          "created": "9 months ago",
          "current": 1,
          "desired": 1,
          "id": "0",
          "kind": "deployment",
          "label": "app=mock-deployment-1; chart=mock-release--99.99.99; heritage=Tiller; release=mock-release-1",
          "name": "mock-deployment-1",
          "namespace": "kube-system",
          "ready": 1,
          "selfLink": "/apis/extensions/v1beta1/namespaces/kube-system/deployments/mock-deployment-1",
        },
        Object {
          "action": <ForwardRef(Delete16)
            className="table-action"
            onClick={[Function]}
          />,
          "apigroup": "apps",
          "available": 1,
          "cluster": "mock-cluster-2",
          "created": "9 months ago",
          "current": 1,
          "desired": 1,
          "id": "1",
          "kind": "deployment",
          "label": "app=mock-deployment-2; chart=mock-release--99.99.99; heritage=Tiller; release=mock-release-2",
          "name": "mock-deployment-2",
          "namespace": "kube-system",
          "ready": 1,
          "selfLink": "/apis/extensions/v1beta1/namespaces/kube-system/deployments/mock-deployment-2",
        },
      ]
    }
    sortRow={[Function]}
    translateWithId={[Function]}
  />
  <ResourceModal
    action=""
    item={Object {}}
    modalOpen={false}
    onClose={[Function]}
  />
</Fragment>
`;

exports[`ResourceTable component header button onClick event should display (node) resource table 1`] = `
<Fragment>
  <div
    className="search--resource-table-header"
  >
    <div>
      <button
        className="search--resource-table-header-button"
        onClick={[Function]}
      >
        <span
          className="linked-resources"
        >
          node(2)
        </span>
        <span
          className="arrow-up"
        >
          ▲
        </span>
      </button>
    </div>
  </div>
  <DataTable
    filterRows={[Function]}
    headers={
      Array [
        Object {
          "header": "Name",
          "key": "name",
        },
        Object {
          "header": "Cluster",
          "key": "cluster",
        },
        Object {
          "header": "Role",
          "key": "role",
        },
        Object {
          "header": "Architecture",
          "key": "architecture",
        },
        Object {
          "header": "OS Image",
          "key": "osImage",
        },
        Object {
          "header": "CPU",
          "key": "cpu",
        },
        Object {
          "header": "Created",
          "key": "created",
        },
        Object {
          "header": "",
          "key": "action",
        },
      ]
    }
    key="node-resource-table"
    locale="en"
    render={[Function]}
    rows={
      Array [
        Object {
          "action": <ForwardRef(Delete16)
            className="table-action"
            onClick={[Function]}
          />,
          "architecture": "amd64",
          "cluster": "mock-cluster-1",
          "cpu": 8,
          "created": "9 months ago",
          "id": "0",
          "kind": "node",
          "label": "beta.kubernetes.io/arch=amd64; beta.kubernetes.io/os=linux; etcd=true; kubernetes.io/hostname=mock-node-1; management=true; master=true; node-role.kubernetes.io/etcd=true; node-role.kubernetes.io/management=true; node-role.kubernetes.io/master=true; node-role.kubernetes.io/proxy=true; node-role.kubernetes.io/va=true; proxy=true; role=master; va=true",
          "name": "mock-node-1",
          "osImage": "Ubuntu 18.04.3 LTS",
          "role": "etcd, management, master, proxy, va",
          "selfLink": "/api/v1/nodes/mock-node-1",
        },
        Object {
          "action": <ForwardRef(Delete16)
            className="table-action"
            onClick={[Function]}
          />,
          "architecture": "amd64",
          "cluster": "mock-cluster-2",
          "cpu": 8,
          "created": "9 months ago",
          "id": "1",
          "kind": "node",
          "label": "beta.kubernetes.io/arch=amd64; beta.kubernetes.io/os=linux; etcd=true; kubernetes.io/hostname=mock-node-2; management=true; master=true; node-role.kubernetes.io/etcd=true; node-role.kubernetes.io/management=true; node-role.kubernetes.io/master=true; node-role.kubernetes.io/proxy=true; node-role.kubernetes.io/va=true; proxy=true; role=master; va=true",
          "name": "mock-node-2",
          "osImage": "Ubuntu 18.04.3 LTS",
          "role": "etcd, management, master, proxy, va",
          "selfLink": "/api/v1/nodes/mock-node-2",
        },
      ]
    }
    sortRow={[Function]}
    translateWithId={[Function]}
  />
  <ResourceModal
    action=""
    item={Object {}}
    modalOpen={false}
    onClose={[Function]}
  />
</Fragment>
`;

exports[`ResourceTable component header button onClick event should display (release) resource table 1`] = `
<Fragment>
  <div
    className="search--resource-table-header"
  >
    <div>
      <button
        className="search--resource-table-header-button"
        onClick={[Function]}
      >
        <span
          className="linked-resources"
        >
          release(2)
        </span>
        <span
          className="arrow-up"
        >
          ▲
        </span>
      </button>
    </div>
  </div>
  <DataTable
    filterRows={[Function]}
    headers={
      Array [
        Object {
          "header": "Name",
          "key": "name",
        },
        Object {
          "header": "Namespace",
          "key": "namespace",
        },
        Object {
          "header": "Cluster",
          "key": "cluster",
        },
        Object {
          "header": "Status",
          "key": "status",
        },
        Object {
          "header": "Chart Name",
          "key": "chartName",
        },
        Object {
          "header": "Chart Version",
          "key": "chartVersion",
        },
        Object {
          "header": "Updated",
          "key": "updated",
        },
        Object {
          "header": "",
          "key": "action",
        },
      ]
    }
    key="release-resource-table"
    locale="en"
    render={[Function]}
    rows={
      Array [
        Object {
          "action": null,
          "chartName": "mock-release-1",
          "chartVersion": "99.99.99",
          "cluster": "mock-cluster-1",
          "id": "0",
          "kind": "release",
          "name": "mock-release-1",
          "namespace": "kube-system",
          "revision": 1,
          "status": "DEPLOYED",
          "updated": "Invalid date",
        },
        Object {
          "action": null,
          "chartName": "mock-release-2",
          "chartVersion": "99.99.99",
          "cluster": "mock-cluster-2",
          "id": "1",
          "kind": "release",
          "name": "mock-release-2",
          "namespace": "kube-system",
          "revision": 1,
          "status": "DEPLOYED",
          "updated": "Invalid date",
        },
      ]
    }
    sortRow={[Function]}
    translateWithId={[Function]}
  />
  <ResourceModal
    action=""
    item={Object {}}
    modalOpen={false}
    onClose={[Function]}
  />
</Fragment>
`;

exports[`ResourceTable component header button onClick event should display (replicaset) resource table 1`] = `
<Fragment>
  <div
    className="search--resource-table-header"
  >
    <div>
      <button
        className="search--resource-table-header-button"
        onClick={[Function]}
      >
        <span
          className="linked-resources"
        >
          replicaset(2)
        </span>
        <span
          className="arrow-up"
        >
          ▲
        </span>
      </button>
    </div>
  </div>
  <DataTable
    filterRows={[Function]}
    headers={
      Array [
        Object {
          "header": "Name",
          "key": "name",
        },
        Object {
          "header": "Namespace",
          "key": "namespace",
        },
        Object {
          "header": "Cluster",
          "key": "cluster",
        },
        Object {
          "header": "Desired",
          "key": "desired",
        },
        Object {
          "header": "Current",
          "key": "current",
        },
        Object {
          "header": "Created",
          "key": "created",
        },
        Object {
          "header": "",
          "key": "action",
        },
      ]
    }
    key="replicaset-resource-table"
    locale="en"
    render={[Function]}
    rows={
      Array [
        Object {
          "action": <ForwardRef(Delete16)
            className="table-action"
            onClick={[Function]}
          />,
          "apigroup": "apps",
          "cluster": "mock-cluster-1",
          "created": "9 months ago",
          "current": 1,
          "desired": 1,
          "id": "0",
          "kind": "replicaset",
          "label": "app=mock-replicaset; chart=mock-release--99.99.99; pod-template-hash=1; release=mock-release-1",
          "name": "mock-replicaset-1",
          "namespace": "kube-system",
          "selfLink": "/apis/extensions/v1beta1/namespaces/kube-system/replicasets/mock-replicaset-1",
        },
        Object {
          "action": <ForwardRef(Delete16)
            className="table-action"
            onClick={[Function]}
          />,
          "apigroup": "apps",
          "cluster": "mock-cluster-2",
          "created": "9 months ago",
          "current": 1,
          "desired": 1,
          "id": "1",
          "kind": "replicaset",
          "label": "app=mock-replicaset; chart=mock-release--99.99.99; pod-template-hash=1; release=mock-release-2",
          "name": "mock-replicaset-2",
          "namespace": "kube-system",
          "selfLink": "/apis/extensions/v1beta1/namespaces/kube-system/replicasets/mock-replicaset-2",
        },
      ]
    }
    sortRow={[Function]}
    translateWithId={[Function]}
  />
  <ResourceModal
    action=""
    item={Object {}}
    modalOpen={false}
    onClose={[Function]}
  />
</Fragment>
`;

exports[`ResourceTable component header button onClick event should display (secret) resource table 1`] = `
<Fragment>
  <div
    className="search--resource-table-header"
  >
    <div>
      <button
        className="search--resource-table-header-button"
        onClick={[Function]}
      >
        <span
          className="linked-resources"
        >
          secret(2)
        </span>
        <span
          className="arrow-up"
        >
          ▲
        </span>
      </button>
    </div>
  </div>
  <DataTable
    filterRows={[Function]}
    headers={
      Array [
        Object {
          "header": "Name",
          "key": "name",
        },
        Object {
          "header": "Namespace",
          "key": "namespace",
        },
        Object {
          "header": "Cluster",
          "key": "cluster",
        },
        Object {
          "header": "Type",
          "key": "type",
        },
        Object {
          "header": "Created",
          "key": "created",
        },
        Object {
          "header": "",
          "key": "action",
        },
      ]
    }
    key="secret-resource-table"
    locale="en"
    render={[Function]}
    rows={
      Array [
        Object {
          "action": <ForwardRef(Delete16)
            className="table-action"
            onClick={[Function]}
          />,
          "cluster": "mock-cluster-1",
          "created": "9 months ago",
          "id": "0",
          "kind": "secret",
          "name": "mock-secret-1",
          "namespace": "kube-system",
          "selfLink": "/api/v1/namespaces/kube-system/secrets/mock-secret-1",
        },
        Object {
          "action": <ForwardRef(Delete16)
            className="table-action"
            onClick={[Function]}
          />,
          "cluster": "mock-cluster-2",
          "created": "9 months ago",
          "id": "1",
          "kind": "secret",
          "name": "mock-secret-2",
          "namespace": "kube-system",
          "selfLink": "/api/v1/namespaces/kube-system/secrets/mock-secret-2",
        },
      ]
    }
    sortRow={[Function]}
    translateWithId={[Function]}
  />
  <ResourceModal
    action=""
    item={Object {}}
    modalOpen={false}
    onClose={[Function]}
  />
</Fragment>
`;

exports[`ResourceTable component header button onClick event should display (service) resource table 1`] = `
<Fragment>
  <div
    className="search--resource-table-header"
  >
    <div>
      <button
        className="search--resource-table-header-button"
        onClick={[Function]}
      >
        <span
          className="linked-resources"
        >
          service(2)
        </span>
        <span
          className="arrow-up"
        >
          ▲
        </span>
      </button>
    </div>
  </div>
  <DataTable
    filterRows={[Function]}
    headers={
      Array [
        Object {
          "header": "Name",
          "key": "name",
        },
        Object {
          "header": "Namespace",
          "key": "namespace",
        },
        Object {
          "header": "Cluster",
          "key": "cluster",
        },
        Object {
          "header": "Created",
          "key": "created",
        },
        Object {
          "header": "",
          "key": "action",
        },
      ]
    }
    key="service-resource-table"
    locale="en"
    render={[Function]}
    rows={
      Array [
        Object {
          "action": <ForwardRef(Delete16)
            className="table-action"
            onClick={[Function]}
          />,
          "cluster": "mock-cluster-1",
          "clusterIP": "None",
          "created": "9 months ago",
          "id": "0",
          "kind": "service",
          "label": "app=mock-service; chart=mock-release--99.99.99; heritage=Tiller; release=mock-release-1",
          "name": "mock-memcached-1",
          "namespace": "kube-system",
          "port": "11211/TCP",
          "selfLink": "/api/v1/namespaces/kube-system/services/mock-memcached-1",
          "type": "ClusterIP",
        },
        Object {
          "action": <ForwardRef(Delete16)
            className="table-action"
            onClick={[Function]}
          />,
          "cluster": "mock-cluster-2",
          "clusterIP": "None",
          "created": "9 months ago",
          "id": "1",
          "kind": "service",
          "label": "app=mock-service; chart=mock-release--99.99.99; heritage=Tiller; release=mock-release-1",
          "name": "mock-memcached-2",
          "namespace": "kube-system",
          "port": "11211/TCP",
          "selfLink": "/api/v1/namespaces/kube-system/services/mock-memcached-2",
          "type": "ClusterIP",
        },
      ]
    }
    sortRow={[Function]}
    translateWithId={[Function]}
  />
  <ResourceModal
    action=""
    item={Object {}}
    modalOpen={false}
    onClose={[Function]}
  />
</Fragment>
`;
